docs/rfc/rfc-018-bls-agg-exploration.md

```diff
@@ -0,0 +1,400 @@
+`G1` and `G2` which are points of the original curve _modulo_ different values.
+Finally, you define a third group `Gt`, where points from `G1` and `G2` satisfy
+the property of bilinearity with `Gt`. In this scheme, the function `e` takes
+as inputs points in `G1` and `G2` and outputs values in `Gt`. Succintly, given
+some point `P` in `G1` and some point `Q` in `G1`, `e(P, Q) = C` where `C` is in `Gt`.
+You can efficiently compute the mapping of points in `G1` and `G2` into `Gt`,
+but you cannot efficiently determine what points were summed and paired to
+produce the value in `Gt`.
+
+Functions are then defined to map digital signatures, messages, and keys into
+and out of points of `G1` or `G2` and signature verification is the process
+of calculating if a set of values representing a message, public key, and digital
+signature produce the same value in `Gt` through `e`.
+
+Signatures can be created as either points in `G1` with public keys being
+created as points in `G2` or vice versa. For the case of BLS12-381, the popular
+curve used, points in `G1` are represented with 48 bytes and points in `G2` are
+represented with 96 bytes. It is up to the implementer of the cryptosystem to
+decide which should be larger, the public keys or the signatures.
+
+BLS signatures rely on pairing-based elliptic-curve cryptography to produce
+various types of signatures. For a more in-depth but still high level discussion
+pairing-based elliptic-curve cryptography, see Vitalik Buterin's post on
+[Exploring Elliptic Curve Pairings][vitalik-curve-post]. For much more in
+depth discussion, see the specific paper on BLS12-381, 
+[Compact Multi-Signatures for Smaller Blockchains][multi-signatures-smaller-blockchains].
+
+### Adoption
+
+BLS signatures have already gained traction within several popular projects.
+
+* Algorand is working on an implementation.
+* [Zcash][zcash-adoption] has adopted BLS12-381 into the protocol.
+* [Ethereum 2.0][eth-2-adoption] has adopted BLS12-381 into the protocol.
+* [Chia Network][chia-adoption] has adopted BLS for signing blocks.
+* [Ostracon][line-ostracon-pr], a fork of Tendermint has adopted BLS for signing blocks.
+
+### What systems may be affected by adding aggregated signatures?
+
+#### Gossip
+
+Gossip could be updated to aggregate vote signatures during a consensus round.
+This appears to be of frankly little utility. Creating an aggregated signature
+incurs overhead, so frequently re-aggregating may incur a significant
+overhead. How costly this is is still subject to further investigation and
+performance testing.
+
+Each validator will still need to receive vote extension data
+from the peer validators in order for consensus to proceed. As a result, any
+advantage gained by aggregating signatures across the vote message will be
+nullified as a result of the addition of vote extensions.
+
+#### Block Creation
+
+When creating a block, the proposer may create a small set of short
```

Comment:
Is "short" necessary here? I can see why it would be desirable, but is it a constraint?

Reply:
Also: A _set_ of multi-signatures?  What are the elements of (the/each) multi-sig in this case? Other validators? Why would there be more than one such set?

(Not saying there isn't a reason, but I don't know what it is)

Reply:
> Is "short" necessary here? I can see why it would be desirable, but is it a constraint?

No, it is not a constraint. You could certainly create a multi-signature that is linear size in the number of signers. Creating a short signature is, imo, one of the most desirable qualities of BLS signatures but not necessarily a requirement of adopting this scheme.

>  A set of multi-signatures? What are the elements of (the/each) multi-sig in this case? Other validators? Why would there be more than one such set?

I believe that this question relates somewhat to a proposed implementation. The section on feasibility, I believe, touches on why this would be a set of multi-signatures. We currently include both the 'yea' and the 'nay' signatures in the block. The 'set' of multi-signatures I was considering was one signature for the yeas and one for the nays. All of the yeas sign over identical bytes and same for the nays. 

This information is currently encoded in the [CommitSig.BlockIDFlag](https://github.com/tendermint/tendermint/blob/851c0dc4f3088723c724cb01c4877ee78c3d6ffc/types/block.go#L589) field.

Reply:
> We currently include both the 'yea' and the 'nay' signatures in the block. The 'set' of multi-signatures I was considering was one signature for the yeas and one for the nays.

I saw that later and thought it might be related to this, but I forgot to come back and add a note. 🙂